src/Analyzers/Core/Analyzers/UseCollectionInitializer/AbstractObjectCreationExpressionAnalyzer.cs

```diff
@@ -24,166 +22,44 @@ internal abstract class AbstractObjectCreationExpressionAnalyzer<
         where TObjectCreationExpressionSyntax : TExpressionSyntax
         where TVariableDeclaratorSyntax : SyntaxNode
     {
-        protected SemanticModel _semanticModel;
-        protected ISyntaxFacts _syntaxFacts;
+        protected UpdateExpressionState<TExpressionSyntax, TStatementSyntax> State;
+
         protected TObjectCreationExpressionSyntax _objectCreationExpression;
         protected bool _analyzeForCollectionExpression;
-        protected CancellationToken _cancellationToken;
-
-        protected TStatementSyntax _containingStatement;
-        private SyntaxNodeOrToken _valuePattern;
-        private ISymbol _initializedSymbol;
 
-        protected AbstractObjectCreationExpressionAnalyzer()
-        {
-        }
+        protected ISyntaxFacts SyntaxFacts => this.State.SyntaxFacts;
+        protected SemanticModel SemanticModel => this.State.SemanticModel;
 
-        protected abstract bool ShouldAnalyze();
-        protected abstract bool TryAddMatches(ArrayBuilder<TMatch> matches);
+        protected abstract bool ShouldAnalyze(CancellationToken cancellationToken);
+        protected abstract bool TryAddMatches(ArrayBuilder<TMatch> matches, CancellationToken cancellationToken);
 
         public void Initialize(
-            SemanticModel semanticModel,
-            ISyntaxFacts syntaxFacts,
+            UpdateExpressionState<TExpressionSyntax, TStatementSyntax> state,
             TObjectCreationExpressionSyntax objectCreationExpression,
-            bool analyzeForCollectionExpression,
-            CancellationToken cancellationToken)
+            bool analyzeForCollectionExpression)
         {
-            _semanticModel = semanticModel;
-            _syntaxFacts = syntaxFacts;
+            State = state;
             _objectCreationExpression = objectCreationExpression;
             _analyzeForCollectionExpression = analyzeForCollectionExpression;
-            _cancellationToken = cancellationToken;
         }
 
         protected void Clear()
         {
-            _semanticModel = null;
-            _syntaxFacts = null;
+            State = default;
             _objectCreationExpression = null;
             _analyzeForCollectionExpression = false;
-            _cancellationToken = default;
-            _containingStatement = null;
-            _valuePattern = default;
-            _initializedSymbol = null;
         }
 
-        protected ImmutableArray<TMatch> AnalyzeWorker()
+        protected ImmutableArray<TMatch> AnalyzeWorker(CancellationToken cancellationToken)
         {
-            if (!ShouldAnalyze())
-                return default;
-
-            _containingStatement = _objectCreationExpression.FirstAncestorOrSelf<TStatementSyntax>();
-            if (_containingStatement == null)
-                return default;
-
-            if (!TryInitializeVariableDeclarationCase() &&
-                !TryInitializeAssignmentCase())
-            {
```

Comment:
because the state-value is immutable (and passed in from teh caller), it became the case that the caller tries to do these initialization checks first.  if successful, it initializes the state-value and then constructs the analyzer.  This prevens the analyzer from needing state that mutates while analysis is performed.